django-backend/fecfiler/reports/models.py

```diff
@@ -0,0 +1,193 @@
+from .form_24.models import Form24
+from .form_99.models import Form99
+import logging
+
+
+logger = logging.getLogger(__name__)
+
+
+class Report(SoftDeleteModel, CommitteeOwnedModel):
+    """Generated model from json schema"""
+
+    id = models.UUIDField(
+        default=uuid.uuid4,
+        editable=False,
+        primary_key=True,
+        serialize=False,
+        unique=True,
+    )
+
+    form_type = models.TextField(null=True, blank=True)
+    report_version = models.TextField(
+        null=True, blank=True
+    )  # fec 1-up version of amendment
+    report_id = models.TextField(null=True, blank=True)  # fec id for report
+    report_code = models.TextField(null=True, blank=True)
+    coverage_from_date = models.DateField(null=True, blank=True)
+    coverage_through_date = models.DateField(null=True, blank=True)
+    treasurer_last_name = models.TextField(null=True, blank=True)
+    treasurer_first_name = models.TextField(null=True, blank=True)
+    treasurer_middle_name = models.TextField(null=True, blank=True)
+    treasurer_prefix = models.TextField(null=True, blank=True)
+    treasurer_suffix = models.TextField(null=True, blank=True)
+    date_signed = models.DateField(null=True, blank=True)
+    calculation_status = models.CharField(max_length=255, null=True, blank=True)
+
+    confirmation_email_1 = models.EmailField(
+        max_length=44, null=True, blank=True, default=None
+    )
+    confirmation_email_2 = models.EmailField(
+        max_length=44, null=True, blank=True, default=None
+    )
+
+    created = models.DateTimeField(auto_now_add=True)
+    updated = models.DateTimeField(auto_now=True)
+
+    upload_submission = models.ForeignKey(
+        "web_services.UploadSubmission",
+        on_delete=models.SET_NULL,
+        null=True,
+        blank=True,
+    )
+    webprint_submission = models.ForeignKey(
+        "web_services.WebPrintSubmission",
+        on_delete=models.SET_NULL,
+        null=True,
+        blank=True,
+    )
+
+    form_3x = models.ForeignKey(Form3X, on_delete=models.CASCADE, null=True, blank=True)
+    form_24 = models.ForeignKey(Form24, on_delete=models.CASCADE, null=True, blank=True)
+    form_99 = models.ForeignKey(Form99, on_delete=models.CASCADE, null=True, blank=True)
+
+    objects = ReportManager()
+
+    def save(self, *args, **kwargs):
+        # Record if the is a create or update operation
+        create_action = self.created is None
+
+        with db_transaction.atomic():
+            super(Report, self).save(*args, **kwargs)
+
+            # Pull forward any loans with non-zero balances along with their
+            # loan guarantors
+            if create_action and self.coverage_through_date:
+                self.pull_forward_loans()
+                self.pull_forward_debts()
+
+    def pull_forward_loans(self):
+        previous_report = self.get_previous_report()
+
+        if previous_report:
+            loans_to_pull_forward = previous_report.transaction_set.filter(
```

Comment:
we probably only want to pull forward on F3X reports for now and not F24 or F99?